skglm/solvers/prox_newton_solver.py

```diff
@@ -0,0 +1,312 @@
+            print(f"Iteration {t + 1}, {ws_size} feats in subpb.")
+
+        # 2) run prox newton on smaller subproblem
+        for epoch in range(max_epochs):
+            _max_pn_cd_epochs = 1 if epoch == 0 else max_pn_cd_epochs
+            pn_tol = 0 if epoch == 0 else tol
+            if is_sparse:
+                _prox_newton_iter_sparse(
+                    X.data, X.indptr, X.indices, Xw, w, y, penalty, ws,
+                    min_pn_cd_epochs, _max_pn_cd_epochs, max_backtrack, pn_tol)
+            else:
+                _prox_newton_iter(
+                    X, Xw, w, y, penalty, ws, min_pn_cd_epochs, _max_pn_cd_epochs,
+                    max_backtrack, pn_tol)
+
+            # 3) do Anderson acceleration on smaller problem
+            w_acc, Xw_acc = accelerator.extrapolate(w, Xw)
+            p_obj = datafit.value(y, w, Xw) + penalty.value(w)
+            p_obj_acc = datafit.value(y, w_acc, Xw_acc) + penalty.value(w_acc)
+
+            if p_obj_acc < p_obj:
+                w[:] = w_acc
+                Xw[:] = Xw_acc
+
+            if epoch % 5 == 0:  # check every 5 epochs, PN epochs are expensive
+                p_obj = datafit.value(y, w[ws], Xw) + penalty.value(w)
+
+                if is_sparse:
+                    grad = construct_grad_sparse(
+                        X.data, X.indptr, X.indices, y, w, Xw, datafit, ws)
+                else:
+                    grad = construct_grad(X, y, w, Xw, datafit, ws)
+                if ws_strategy == "subdiff":
+                    opt_ws = penalty.subdiff_distance(w, grad, ws)
+                elif ws_strategy == "fixpoint":
+                    opt_ws = dist_fix_point(w, grad, datafit, penalty, ws)
+
+                stop_crit_in = np.max(opt_ws)
+                if max(verbose - 1, 0):
+                    print(f"Epoch {epoch + 1}, objective {p_obj:.10f}, "
+                          f"stopping crit {stop_crit_in:.2e}")
+                if ws_size == n_features:
+                    if stop_crit_in <= tol:
+                        break
+                else:
+                    if stop_crit_in < 0.3 * stop_crit:
+                        if max(verbose - 1, 0):
+                            print("Early exit")
+                        break
+        obj_out.append(p_obj)
+    return w, np.array(obj_out), stop_crit
+
+
+@njit
+def _prox_newton_iter(
+    X, Xw, w, y, penalty, ws, min_cd_epochs, max_cd_epochs, max_backtrack, tol
+):
+    n_samples, ws_size = X.shape[0], ws.shape[0]
+
+    hessian_diag = np.zeros(n_samples)  # hessian = X^T D X, with D = diag(f_i'')
+    grad_datafit = np.zeros(n_samples)  # gradient of F(Xw)
+
+    lc = np.zeros(ws_size)  # weighted Lipschitz constants
+    bias = np.zeros(ws_size)
+
+    for i in range(n_samples):
+        prob = 1. / (1. + np.exp(y[i] * Xw[i]))
+        # this supposes that the datafit is scaled by n_samples
+        hessian_diag[i] = prob * (1. - prob) / n_samples
+        grad_datafit[i] = -y[i] * prob / n_samples
```

Comment:
Why do you hard code the ``grad`` here? Use ``gradient_scalar`` of the ``datafit`` instead

Reply:
I'm not sure it computes the same thing. `gradient_scalar` is the grad of the datafit w.r.t. to `w`, while I'm computing the gradient w.r.t. `Xw`. I haven't checked the computations. 

Anyway, I think it's efficient to do it this way since we reuse `prob` to compute `hessian_diag` and `grad_datafit`. WDYT?

Reply:
This is how I view the grad considering the datafit: $F(Xw)$ where $F(u) = \frac{1}{n} \sum_i f_i(u_i)$

$\nabla (F(Xw)) = X^T \\nabla F(Xw)$

such that $\\nabla F(u) = (f_i'(u_i))_{1 \leq i \leq n}$

Reply:
IMO, computing the hessian, weighted dot, ... should be handled by the datafit itself and the solver shouldn't interfere/intervene with it at all.

I believe that a PN solver is a just ``cd_solver`` with a different descent direction/step.

Reply:
When calling `gradient_scalar` with a loop over the features, you obtain the gradient of the datafit w.r.t. to the features. Here I'm looping over the samples to obtain the gradient of `F(Xw)`. It's not the same quantity.

As for the design, I agree, but let's keep this PR simple and refactor to handle other datafits in a later PR.

Reply:
@Badr-MOUFAD  PN requires quantities that are different from the ones that `cd_solver` needs, and the optimization algorithm is quite different.